{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1) Downloaded the PDC 1, PDC 5, PDC 6 Gene of Saccharomyces cerevisiae"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDC1 gene sequence\n",
      "Downloaded PDC1 gene sequence successfully\n",
      "Downloading PDC5 gene sequence\n",
      "Downloaded PDC5 gene sequence successfully\n",
      "Downloading PDC6 gene sequence\n",
      "Downloaded PDC6 gene sequence successfully\n"
     ]
    }
   ],
   "source": [
    "# Downloading the PDC Gene's (1,5,6)\n",
    "from Bio import SeqIO, Entrez\n",
    "\n",
    "Entrez.email = \"sameer22439@iiitd.ac.in\"\n",
    "\n",
    "# Function to download gene sequence from NCBI\n",
    "def download_gene_sequence(output_file, accession_number, start_pos, end_pos, reverse_complement=False):\n",
    "    \n",
    "    # Fetches the gene sequence from NCBI using accession number, start, and end positions. Optionally takes the reverse complement if specified.\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"fasta\", seq_start=start_pos, seq_stop=end_pos)\n",
    "\n",
    "    gene_record = SeqIO.read(handle, \"fasta\")\n",
    "\n",
    "    handle.close()\n",
    "\n",
    "    if reverse_complement: # If reverse complement is specified, reverse complement the sequence\n",
    "        gene_record.seq = gene_record.seq.reverse_complement()\n",
    "\n",
    "    with open(output_file, \"w\") as file:\n",
    "        SeqIO.write(gene_record, file, \"fasta\")\n",
    "\n",
    "    return gene_record\n",
    "\n",
    "# PDC 1 requires reverse complement \n",
    "accession_number_pdc1 = \"NC_001144.5\" # Accession number of PDC1 gene\n",
    "start_position_pdc1 = 232390 # Start position of PDC1 gene\n",
    "end_position_pdc1 = 234081 # End position of PDC1 gene\n",
    "\n",
    "print(\"Downloading PDC1 gene sequence\")\n",
    "gene_record_pdc1 = download_gene_sequence(\"PDC1.fsa\", accession_number_pdc1, start_position_pdc1, end_position_pdc1, True)\n",
    "print(\"Downloaded PDC1 gene sequence successfully\")\n",
    "\n",
    "# PDC 5 doesn't requires reverse complement\n",
    "accession_number_pdc5 = \"NC_001144.5\" # Accession number of PDC5 gene\n",
    "start_position_pdc5 = 410723 # Start position of PDC5 gene\n",
    "end_position_pdc5 = 412414 # End position of PDC5 gene\n",
    "\n",
    "print(\"Downloading PDC5 gene sequence\")\n",
    "gene_record_pdc5 = download_gene_sequence(\"PDC5.fsa\", accession_number_pdc5, start_position_pdc5, end_position_pdc5)\n",
    "print(\"Downloaded PDC5 gene sequence successfully\")\n",
    "\n",
    "# PDC 6 requires reverse complement\n",
    "accession_number_pdc6 = \"NC_001139.9\" # Accession number of PDC6 gene\n",
    "start_position_pdc6 = 651290 # Start position of PDC6 gene\n",
    "end_position_pdc6 = 652981 # End position of PDC6 gene\n",
    "\n",
    "print(\"Downloading PDC6 gene sequence\")\n",
    "gene_record_pdc6 = download_gene_sequence(\"PDC6.fsa\", accession_number_pdc6, start_position_pdc6, end_position_pdc6, True)\n",
    "print(\"Downloaded PDC6 gene sequence successfully\")\n",
    "\n",
    "# print(gene_record_pdc1.seq)\n",
    "# print(gene_record_pdc5)\n",
    "# print(gene_record_pdc6)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2) Performing Blast to compare PDC Gene Sequence across different Saccharomyces cerevisiae strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading BLAST file for PDC1 gene\n",
      "BLASTing the sequence against the NCBI database\n",
      "BLASTing completed successfully\n",
      "BLAST results saved to PDC1_blast.xml\n",
      "Downloaded BLAST file for PDC1 gene successfully\n",
      "Parsed BLAST results for PDC1 gene successfully\n",
      "---------------------------------------------\n",
      "Downloading BLAST file for PDC5 gene\n",
      "BLASTing the sequence against the NCBI database\n",
      "BLASTing completed successfully\n",
      "BLAST results saved to PDC5_blast.xml\n",
      "Downloaded BLAST file for PDC5 gene successfully\n",
      "Parsed BLAST results for PDC5 gene successfully\n",
      "---------------------------------------------\n",
      "Downloading BLAST file for PDC6 gene\n",
      "BLASTing the sequence against the NCBI database\n",
      "BLASTing completed successfully\n",
      "BLAST results saved to PDC6_blast.xml\n",
      "Downloaded BLAST file for PDC6 gene successfully\n",
      "Parsed BLAST results for PDC6 gene successfully\n",
      "---------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio import SeqIO, Entrez\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Function to remove gaps and ambigious bases from the sequence\n",
    "def removing_gaps_and_ambigious_bases(sequence):\n",
    "    # Removes gaps and ambigious bases from the sequence\n",
    "    final_sequence = str(sequence).replace(\"-\", \"\").replace(\"N\", \"\")\n",
    "\n",
    "    while len(final_sequence) % 3 != 0:\n",
    "        final_sequence = final_sequence + \"N\"\n",
    "    \n",
    "    return Seq(final_sequence)\n",
    "\n",
    "# Function to download BLAST results for the given sequence\n",
    "def download_blast_file(sequence, output_file):\n",
    "    # Removes gaps and ambigious bases from the sequence\n",
    "    sequence = removing_gaps_and_ambigious_bases(sequence)\n",
    "\n",
    "    # BLAST the sequence against the NCBI database\n",
    "    print(\"BLASTing the sequence against the NCBI database\")\n",
    "    result_handle = NCBIWWW.qblast(\"blastn\", \"nr\", sequence, entrez_query='txid4932[Organism]')\n",
    "    print(\"BLASTing completed successfully\")\n",
    "    \n",
    "    with open(output_file, \"w\") as file:\n",
    "        file.write(result_handle.read())\n",
    "    \n",
    "    print(\"BLAST results saved to\", output_file)\n",
    "\n",
    "    result_handle.close()\n",
    "\n",
    "# Function to parse the BLAST results and save the scores to a text file\n",
    "def parse_blast_results(blast_file, output_file):\n",
    "    with open(blast_file, \"r\") as result_handle, open(output_file, \"w\") as output_text:\n",
    "        blast_records = NCBIXML.parse(result_handle)\n",
    "\n",
    "        for blast_record in blast_records:\n",
    "            for alignment in blast_record.alignments:\n",
    "                output_text.write(f\"Alignment title: {alignment.title}\\n\\n\")\n",
    "                for hsp in alignment.hsps:\n",
    "                    output_text.write(f\"Score:   {hsp.score}\\n\")\n",
    "                    output_text.write(f\"E-value: {hsp.expect}\\n\")\n",
    "                    output_text.write(f\"Query:   {hsp.query}\\n\")\n",
    "                    output_text.write(f\"Match:   {hsp.match}\\n\")\n",
    "                    subject_seq = str(hsp.sbjct).replace(\"-\", \"\").replace(\"N\", \"\")\n",
    "                    output_text.write(f\"Subject: {subject_seq}\\n\")\n",
    "                    \n",
    "                output_text.write(\"-------------------------------------------------------------\\n\\n\")\n",
    "            \n",
    "\n",
    "\n",
    "# Download the BLAST results for PDC1 gene\n",
    "print(\"Downloading BLAST file for PDC1 gene\")\n",
    "download_blast_file(gene_record_pdc1.seq, \"PDC1_blast.xml\")\n",
    "print(\"Downloaded BLAST file for PDC1 gene successfully\")\n",
    "parse_blast_results(\"PDC1_blast.xml\", \"PDC1_blast_score.txt\")\n",
    "print(\"Parsed BLAST results for PDC1 gene successfully\")\n",
    "print(\"---------------------------------------------\")\n",
    "\n",
    "# Download the BLAST results for PDC5 gene\n",
    "print(\"Downloading BLAST file for PDC5 gene\")\n",
    "download_blast_file(gene_record_pdc5.seq, \"PDC5_blast.xml\")\n",
    "print(\"Downloaded BLAST file for PDC5 gene successfully\")\n",
    "parse_blast_results(\"PDC5_blast.xml\", \"PDC5_blast_score.txt\")\n",
    "print(\"Parsed BLAST results for PDC5 gene successfully\")\n",
    "print(\"---------------------------------------------\")\n",
    "\n",
    "# Download the BLAST results for PDC6 gene\n",
    "print(\"Downloading BLAST file for PDC6 gene\")\n",
    "download_blast_file(gene_record_pdc6.seq, \"PDC6_blast.xml\")\n",
    "print(\"Downloaded BLAST file for PDC6 gene successfully\")\n",
    "parse_blast_results(\"PDC6_blast.xml\", \"PDC6_blast_score.txt\")\n",
    "print(\"Parsed BLAST results for PDC6 gene successfully\")\n",
    "print(\"---------------------------------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding SNPs in PDC1 gene\n",
      "Finding SNPs in PDC5 gene\n",
      "Finding SNPs in PDC6 gene\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio import SeqIO, Entrez\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Function to find SNPs in the BLAST results\n",
    "def finding_SNP_in_blast_results(blast_file, output_file):\n",
    "    with open(blast_file, \"r\") as result_handle, open(output_file, \"w\") as output_text:\n",
    "        blast_records = NCBIXML.parse(result_handle)\n",
    "\n",
    "        for blast_record in blast_records:\n",
    "            for alignment in blast_record.alignments:\n",
    "                output_text.write(f\"Alignment title: {alignment.title}\\n\\n\")\n",
    "                for hsp in alignment.hsps:\n",
    "                    output_text.write(f\"Query:   {hsp.query}\\n\")\n",
    "                    output_text.write(f\"Match:   {hsp.match}\\n\")\n",
    "                    subject_seq = str(hsp.sbjct).replace(\"-\", \"\").replace(\"N\", \"\")\n",
    "                    output_text.write(f\"Subject: {subject_seq}\\n\")\n",
    "\n",
    "                    flag = 1\n",
    "\n",
    "                    for i in range(len(hsp.query)):\n",
    "                        if (hsp.query[i] != hsp.sbjct[i]):\n",
    "                            if (hsp.query[i-1] == hsp.sbjct[i-1] and hsp.query[i+1] == hsp.sbjct[i+1]):\n",
    "                                output_text.write(f\"SNP found at position {i+1}: {hsp.query[i]} ---> {hsp.sbjct[i]}\\n\")\n",
    "                                flag = 0\n",
    "                \n",
    "                    if flag == 1:\n",
    "                        output_text.write(\"No SNP found\\n\")\n",
    "\n",
    "\n",
    "                output_text.write(\"-------------------------------------------------------------\\n\")\n",
    "\n",
    "                            \n",
    "print(\"Finding SNPs in PDC1 gene\")\n",
    "finding_SNP_in_blast_results(\"PDC1_blast.xml\", \"PDC1_SNP.txt\")\n",
    "print(\"Finding SNPs in PDC5 gene\")\n",
    "finding_SNP_in_blast_results(\"PDC5_blast.xml\", \"PDC5_SNP.txt\")\n",
    "print(\"Finding SNPs in PDC6 gene\")\n",
    "finding_SNP_in_blast_results(\"PDC6_blast.xml\", \"PDC6_SNP.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3) Translation of PDC genes and their SNP containing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Translating PDC1 gene and it's strains\n",
      "Translating PDC5 gene and it's strains\n",
      "Translating PDC6 gene and it's strains\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio import SeqIO, Entrez\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Function to translate the strains and query in the BLAST results\n",
    "def translating_the_strains_and_query_in_blast_results(blast_file, output_file):\n",
    "    with open(blast_file, \"r\") as result_handle, open(output_file, \"w\") as output_text:\n",
    "        blast_records = NCBIXML.parse(result_handle)\n",
    "        gene_sequence = None\n",
    "        for blast_record in blast_records:\n",
    "            for alignment in blast_record.alignments:\n",
    "                output_text.write(f\"Alignment title: {alignment.title}\\n\\n\")\n",
    "                for hsp in alignment.hsps:\n",
    "                    gene_sequence = hsp.query\n",
    "                    subject_seq = str(hsp.sbjct).replace(\"-\", \"\").replace(\"N\", \"\")\n",
    "                    protein_seq = Seq(subject_seq).translate()\n",
    "                    output_text.write(f\"Subject : {subject_seq}\\n\")\n",
    "                    output_text.write(f\"Protein : {protein_seq}\\n\\n\")\n",
    "\n",
    "                output_text.write(\"-------------------------------------------------------------\\n\")\n",
    "        \n",
    "        protein_sequence_of_query = str(Seq(gene_sequence).translate())\n",
    "        output_text.write(\"Query :   \" + str(gene_sequence) + \"\\n\")\n",
    "        output_text.write(\"Protein : \" + protein_sequence_of_query)\n",
    "\n",
    "\n",
    "print(\"Translating PDC1 gene and it\\'s strains\")\n",
    "translating_the_strains_and_query_in_blast_results(\"PDC1_blast.xml\", \"PDC1_protein.txt\")\n",
    "print(\"Translating PDC5 gene and it\\'s strains\")\n",
    "translating_the_strains_and_query_in_blast_results(\"PDC5_blast.xml\", \"PDC5_protein.txt\")\n",
    "print(\"Translating PDC6 gene and it\\'s strains\")\n",
    "translating_the_strains_and_query_in_blast_results(\"PDC6_blast.xml\", \"PDC6_protein.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4) Done this question using online software (http://genetics.bwh.harvard.edu/pph2/index.shtml) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Links given below are the SNP's in the Query Protein Sequence and it's strains returned during BLASTN process for 3 different genes of Saccharomyces cerevisiae, i.e. PDC 1, PDC 5 and PDC 6\n",
    "\n",
    "\n",
    "session id : 0ee7cc0802f7d10ef08cc68d35b7ead9646c9924\n",
    "\n",
    "link 1 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644046.html : PDC 1\n",
    "\n",
    "link 2 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644047.html : PDC 1\n",
    "\n",
    "link 3 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644048.html : PDC 1\n",
    "\n",
    "link 4 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644049.html : PDC 1\n",
    "\n",
    "link 5 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644050.html : PDC 1\n",
    "\n",
    "link 6 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644054.html : PDC 5\n",
    "\n",
    "link 7 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644057.html : PDC 5\n",
    "\n",
    "link 8 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644058.html : PDC 5\n",
    "\n",
    "link 9 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644060.html : PDC 5\n",
    "\n",
    "link 10 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9644061.html : PDC 5\n",
    "\n",
    "link 11 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9645333.html : PDC 6\n",
    "\n",
    "link 12 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9645338.html : PDC 6\n",
    "\n",
    "link 13 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9645353.html : PDC 6\n",
    "\n",
    "link 14 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9645357.html : PDC 6\n",
    "\n",
    "link 15 : http://genetics.bwh.harvard.edu/ggi/pph2/0ee7cc0802f7d10ef08cc68d35b7ead9646c9924/9645373.html : PDC 6"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
