{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Downloading Fasta files\n",
    "#### Taken help from the Tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "def download_fasta_sequence(email, accession_number, output_file):\n",
    "    Entrez.email = email\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"fasta\", retmode=\"text\")\n",
    "    print(handle.url)\n",
    "    sequence = handle.read()\n",
    "    with open(output_file, \"w\") as f:\n",
    "        f.write(sequence)\n",
    "\n",
    "# Replace with your email, accession number, and output file name\n",
    "email = \"sameer22439@iiitd.ac.in\"\n",
    "accession_number = [\"NC_001133.9\", \"NC_001134.8\", \"NC_001135.5\", \"NC_001136.10\", \"NC_001137.3\", \"NC_001138.5\", \"NC_001139.9\", \"NC_001140.6\", \"NC_001141.2\", \"NC_001142.9\", \"NC_001143.9\", \"NC_001144.5\", \"NC_001145.3\", \"NC_001146.8\", \"NC_001147.6\", \"NC_001148.4\", \"NC_001224.1\"]  #List of Accession Number\n",
    "\n",
    "\n",
    "for i in range (17):\n",
    "    if (i!=16):\n",
    "        output_file = f\"output_{i+1}.fasta\"\n",
    "        download_fasta_sequence(email, accession_number[i], output_file)\n",
    "    else:\n",
    "        output_file = f\"output_MT.fasta\"\n",
    "        download_fasta_sequence(email, accession_number[i], output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Method 1: Using ARS = 5'-WTTTATRTTTW-3' (total 8 possible cases for this method) as W = A or T; R = A or G; there are 2 possible cases of Nucleotide at 3 position therefore 8 cases\n",
    "#### Taken help from the documentation available on YeastGenome Site = \"https://www.yeastgenome.org/locus/S000118318\"\n",
    "#### Taken help from tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "import re\n",
    "\n",
    "def ori_in_sequences(ithChromosome):\n",
    "    output_file = f\"output_{ithChromosome}.fasta\"\n",
    "\n",
    "    with open(output_file, 'r') as file:  # Open file in read mode\n",
    "        lines = file.readlines()\n",
    "\n",
    "    genome_sequence = ''.join(lines[1:])  # Concatenate all lines except the first one (assuming first line is header)\n",
    "    genome_sequence = genome_sequence.replace('\\n', '')  # Remove newline characters\n",
    "\n",
    "\n",
    "    pattern_for_ARS = re.compile(r'(A|T)TTTAT(A|G)TTT(A|T)')\n",
    "    indexes_of_ORI = []\n",
    "\n",
    "    for ori_found_in_sequence in pattern_for_ARS.finditer(genome_sequence):\n",
    "        indexes_of_ORI.append(ori_found_in_sequence.start())            \n",
    "\n",
    "    \n",
    "    print(f\"Chromosome {ithChromosome} of Baker's Yeast (Saccharomyces cerevisiae) : \")\n",
    "    count = 0\n",
    "    for i in indexes_of_ORI:\n",
    "        count += 1\n",
    "        print(f\"ORI - {count} : Start Index - {i}; End Index - {i+11}\")\n",
    "        print(genome_sequence[i:i+12])\n",
    "\n",
    "    if (count == 0):\n",
    "        print(f\"There is no ORI in the Chromosome {ithChromosome} of Baker's Yeast (Saccharomyces cerevisiae)\")\n",
    "    \n",
    "    return count\n",
    "\n",
    "def ori_in_sequences_in_Mitochondria():\n",
    "    output_file = f\"output_MT.fasta\"\n",
    "\n",
    "    with open(output_file, 'r') as file:  # Open file in read mode\n",
    "        lines = file.readlines()\n",
    "\n",
    "    genome_sequence = ''.join(lines[1:])  # Concatenate all lines except the first one (assuming first line is header)\n",
    "    genome_sequence = genome_sequence.replace('\\n', '')  # Remove newline characters\n",
    "\n",
    "\n",
    "    pattern_for_ARS = re.compile(r'(A|T)TTTAT(A|G)TTT(A|T)')\n",
    "    indexes_of_ORI = []\n",
    "\n",
    "    for ori_found_in_sequence in pattern_for_ARS.finditer(genome_sequence):\n",
    "        indexes_of_ORI.append(ori_found_in_sequence.start())            \n",
    "\n",
    "    \n",
    "    print(f\"Mitochondria of Baker's Yeast (Saccharomyces cerevisiae) : \")\n",
    "    count = 0\n",
    "    for i in indexes_of_ORI:\n",
    "        count += 1\n",
    "        print(f\"ORI - {count} : Start Index - {i}; End Index - {i+11}\")\n",
    "        print(genome_sequence[i:i+12])\n",
    "\n",
    "    if (count == 0):\n",
    "        print(\"There is no ORI in Mitochondria of Baker's Yeast (Saccharomyces cerevisiae)\")\n",
    "        \n",
    "    return count\n",
    "\n",
    "def main():\n",
    "    count = 0\n",
    "    for i in range(1,17):\n",
    "        count += ori_in_sequences(i)\n",
    "        # print(count)\n",
    "        print(\"\\n\\n\")\n",
    "    count += ori_in_sequences_in_Mitochondria()\n",
    "    print(f\"\\n\\nThere are total {count} ORI's in the total genome of Baker's Yeast (Saccharomyces cerevisiae)\")\n",
    "\n",
    "\n",
    "main()\n",
    "       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Method 2: Using ARS = 5'-WTTTAYRTTTW-3' (total 16 cases for this method) as W = A or T; Y = C or T; R = A or G; there are 2 possible cases of Nucleotide at 4 position therefore 16 cases\n",
    "### More Precise than Method 1\n",
    "#### Taken help from the documentation available on YeastGenome Site = \"https://www.yeastgenome.org/locus/S000118318\"\n",
    "#### Taken help from tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "import re\n",
    "\n",
    "def ori_in_sequences(ithChromosome):\n",
    "    output_file = f\"output_{ithChromosome}.fasta\"\n",
    "\n",
    "    with open(output_file, 'r') as file:  # Open file in read mode\n",
    "        lines = file.readlines()\n",
    "\n",
    "    genome_sequence = ''.join(lines[1:])  # Concatenate all lines except the first one (assuming first line is header)\n",
    "    genome_sequence = genome_sequence.replace('\\n', '')  # Remove newline characters\n",
    "\n",
    "\n",
    "    pattern_for_ARS = re.compile(r'(A|T)TTTA(C|T)(A|G)TTT(A|T)')\n",
    "    indexes_of_ORI = []\n",
    "\n",
    "    for ori_found_in_sequence in pattern_for_ARS.finditer(genome_sequence):\n",
    "        indexes_of_ORI.append(ori_found_in_sequence.start())            \n",
    "\n",
    "    \n",
    "    print(f\"Chromosome {ithChromosome} of Baker's Yeast (Saccharomyces cerevisiae) : \")\n",
    "    count = 0\n",
    "    for i in indexes_of_ORI:\n",
    "        count += 1\n",
    "        print(f\"ORI - {count} : Start Index - {i}; End Index - {i+11}\")\n",
    "        print(genome_sequence[i:i+12])\n",
    "\n",
    "    if (count == 0):\n",
    "        print(f\"There is no ORI in the Chromosome {ithChromosome} of Baker's Yeast (Saccharomyces cerevisiae)\")\n",
    "    \n",
    "    return count\n",
    "\n",
    "def ori_in_sequences_in_Mitochondria():\n",
    "    output_file = f\"output_MT.fasta\"\n",
    "\n",
    "    with open(output_file, 'r') as file:  # Open file in read mode\n",
    "        lines = file.readlines()\n",
    "\n",
    "    genome_sequence = ''.join(lines[1:])  # Concatenate all lines except the first one (assuming first line is header)\n",
    "    genome_sequence = genome_sequence.replace('\\n', '')  # Remove newline characters\n",
    "\n",
    "\n",
    "    pattern_for_ARS = re.compile(r'(A|T)TTTA(C|T)(A|G)TTT(A|T)')\n",
    "    indexes_of_ORI = []\n",
    "\n",
    "    for ori_found_in_sequence in pattern_for_ARS.finditer(genome_sequence):\n",
    "        indexes_of_ORI.append(ori_found_in_sequence.start())            \n",
    "\n",
    "    \n",
    "    print(f\"Mitochondria of Baker's Yeast (Saccharomyces cerevisiae) : \")\n",
    "    count = 0\n",
    "    for i in indexes_of_ORI:\n",
    "        count += 1\n",
    "        print(f\"ORI - {count} : Start Index - {i}; End Index - {i+11}\")\n",
    "        print(genome_sequence[i:i+12])\n",
    "    \n",
    "    if (count == 0):\n",
    "        print(\"There is no ORI in Mitochondria of Baker's Yeast (Saccharomyces cerevisiae)\")\n",
    "\n",
    "    return count\n",
    "\n",
    "def main():\n",
    "    count = 0\n",
    "    for i in range(1,17):\n",
    "        count += ori_in_sequences(i)\n",
    "        # print(count)\n",
    "        print(\"\\n\\n\")\n",
    "    count += ori_in_sequences_in_Mitochondria()\n",
    "    print(f\"\\n\\nThere are total {count} ORI's in the total genome of Baker's Yeast (Saccharomyces cerevisiae)\")\n",
    "\n",
    "\n",
    "main()\n",
    "       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Method 3: Using AT rich sites for each chromosome, this code is user centric He/She must know which window size the user has to take for particular chromosome and particular threshold percentage of AT content in the window provided by the user.\n",
    "#### Generally for example I have taken 100 as window size and 80% as threshold AT percentage in the window. (*It will just give the general idea of ORI and it might give wrong answer; so see the answers given below for more precise answer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_AT_percentage(sequence):\n",
    "    total_length = len(sequence)\n",
    "    AT_count = 0\n",
    "    for nucleotide in sequence:\n",
    "        if nucleotide.upper() == 'A' or nucleotide.upper() == 'T':\n",
    "            AT_count += 1\n",
    "    return (AT_count / total_length) * 100\n",
    "\n",
    "\n",
    "def main(output_file, ithChromosome):\n",
    "    # Read the DNA sequence from the file\n",
    "    with open(output_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    sequence = ''.join(lines[1:])  # Concatenate all lines except the first one (assuming first line is header)\n",
    "    sequence = sequence.replace('\\n', '')  # Remove newline characters\n",
    "\n",
    "    # Input band size\n",
    "    band_size = int(input(f\"Enter the band size (number of nucleotides per band) you want to take for Chromosome No. {ithChromosome}: \"))\n",
    "    threshold = float(input(f\"Enter the AT threshold percentage you want to take for Chromosome No. {ithChromosome}: \"))\n",
    "\n",
    "    # band_size = 100\n",
    "    # threshold = 80\n",
    "\n",
    "\n",
    "    # Check AT% in every band\n",
    "    current_band_start = 0\n",
    "    count = 1\n",
    "    for i in range(0, len(sequence), band_size):\n",
    "        band = sequence[i:i+band_size]\n",
    "        AT_percentage = calculate_AT_percentage(band)\n",
    "        if AT_percentage >= threshold:\n",
    "            print(f\"ORI {count} of {ithChromosome} Chromosome: Band {i//band_size + 1}, Actual_AT%: {AT_percentage:.2f}%, Start: {current_band_start}, End: {i}\")\n",
    "            count += 1\n",
    "        current_band_start = i + 1\n",
    "\n",
    "\n",
    "\n",
    "for i in range (1,18):\n",
    "    output_file = f\"output_{i}.fasta\"\n",
    "    \n",
    "    if (i != 17):\n",
    "        main(output_file, i)\n",
    "    else:\n",
    "        main(\"output_MT.fasta\", 17)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Method 4: Directly Extracting from GeneBank by using accession number\n",
    "#### It will always give correct answer as I have directly accessed the data present in the GeneBank to find the ORI's."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Set your email address (required by NCBI)\n",
    "Entrez.email = \"sameer22439@iiitd.ac.in\"\n",
    "\n",
    "def fetch_replication_origins(accession_number, output_file):\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"gb\", retmode=\"text\")\n",
    "    record = handle.read()\n",
    "    handle.close()\n",
    "    \n",
    "    with open(output_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    sequence = ''.join(lines[1:])  # Concatenate all lines except the first one (assuming first line is header)\n",
    "    sequence = sequence.replace('\\n', '')  # Remove newline characters\n",
    "\n",
    "    count = 1\n",
    "\n",
    "    # replication_origins = []\n",
    "    # Parse the GenBank record\n",
    "    for line in record.split(\"\\n\"):\n",
    "        if line.startswith(\"ORIGIN\"):\n",
    "            break\n",
    "        if \"rep_origin\" in line:\n",
    "            # Extract replication origin information\n",
    "            parts = line.strip().split()\n",
    "            start, end = parts[-1].split(\"..\")\n",
    "            # print(parts)\n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "            print(f\"ORI - {count} : Start Index - {start} ; End Index - {end}\")\n",
    "            print(sequence[start-1:end])\n",
    "            count += 1\n",
    "    if count == 1 :\n",
    "        print(\"There is no ORI in this Chromosome\")\n",
    "\n",
    "def generate_complementary_strand(sequence):\n",
    "    complementary_sequence = \"\"\n",
    "    for nucleotide in sequence:\n",
    "        if nucleotide == \"A\":\n",
    "            complementary_sequence += \"T\"\n",
    "        elif nucleotide == \"T\":\n",
    "            complementary_sequence += \"A\"\n",
    "        elif nucleotide == \"C\":\n",
    "            complementary_sequence += \"G\"\n",
    "        elif nucleotide == \"G\":\n",
    "            complementary_sequence += \"C\"\n",
    "        else:\n",
    "            complementary_sequence += nucleotide  # Preserve other characters\n",
    "    return complementary_sequence\n",
    "\n",
    "def fetch_replication_origin_of_MT(accession_number, output_file):\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"gb\", retmode=\"text\")\n",
    "    record = handle.read()\n",
    "    handle.close()\n",
    "\n",
    "    with open(output_file, 'r') as file:  # Open file in read mode\n",
    "        lines = file.readlines()\n",
    "\n",
    "    sequence = ''.join(lines[1:])  # Concatenate all lines except the first one (assuming first line is header)\n",
    "    sequence = sequence.replace('\\n', '')  # Remove newline characters\n",
    "\n",
    "    count = 1\n",
    "    # Parse the GenBank record\n",
    "    for line in record.split(\"\\n\"):\n",
    "        if line.startswith(\"ORIGIN\"):\n",
    "            break\n",
    "        if \"rep_origin\" in line:\n",
    "            parts = line.split('(')[-1].split(')')[0].split('..')\n",
    "            if 'complement' in line:\n",
    "                start = (int(parts[0]))\n",
    "                end = (int(parts[1]))\n",
    "                print(f\"ORI - {count} : Start Index - {start} ; End Index - {end}\")\n",
    "                # print(sequence[start-1:end])\n",
    "                complementary_strand = generate_complementary_strand(sequence[start-1:end])\n",
    "                print(\"It is the complementary strand with the indexes as per the template strand: \" + complementary_strand)\n",
    "                count += 1\n",
    "            else:\n",
    "                nums = [int(x) for x in line.split() if x.isdigit()]\n",
    "                if len(nums) == 1:\n",
    "                    start = int(nums[0])\n",
    "                    end = int(nums[0])\n",
    "                elif len(nums) == 2:\n",
    "                    start = int(nums[0])\n",
    "                    end = int(nums[1])\n",
    "                print(f\"ORI - {count} : Start Index - {start} ; End Index - {end}\")\n",
    "                print(sequence[start-1:end])\n",
    "                count += 1\n",
    "    if count == 1 :\n",
    "        print(\"There is no ORI in Mitochondria\")\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "def main():\n",
    "    # Accession number for baker's yeast chromosome \n",
    "\n",
    "    for i in range(16):\n",
    "        accession_number = [\"NC_001133.9\", \"NC_001134.8\", \"NC_001135.5\", \"NC_001136.10\", \"NC_001137.3\", \"NC_001138.5\", \"NC_001139.9\", \"NC_001140.6\", \"NC_001141.2\", \"NC_001142.9\", \"NC_001143.9\", \"NC_001144.5\", \"NC_001145.3\", \"NC_001146.8\", \"NC_001147.6\", \"NC_001148.4\", \"NC_001224.1\"]  #List of Accession Number\n",
    "        print(f\"Replication Origins for Baker's Yeast Chromosome {i+1}:\")\n",
    "        fetch_replication_origins(accession_number[i], f\"output_{i+1}.fasta\")\n",
    "        print(\"\\n\\n\")\n",
    "\n",
    "    print(f\"Replication Origins for Baker's Yeast Mitochondria :\")\n",
    "    fetch_replication_origin_of_MT(\"NC_001224.1\", \"output_MT.fasta\")\n",
    "\n",
    "\n",
    "main()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
